runtime/commonMain/src/kotlinx/serialization/internal/NullableSerializer.kt

```diff
@@ -5,50 +5,64 @@
 package kotlinx.serialization.internal
 
 import kotlinx.serialization.*
-import kotlin.reflect.KClass
 
-fun <T : Any> makeNullable(actualSerializer: KSerializer<T>): KSerializer<T?> = NullableSerializer(actualSerializer)
 
-class NullableSerializer<T : Any>(public val actualSerializer: KSerializer<T>) : KSerializer<T?> {
-    private class SerialDescriptorForNullable(val original: SerialDescriptor): SerialDescriptor by original {
-        override val isNullable: Boolean
-            get() = true
-
-        override fun equals(other: Any?): Boolean {
-            if (this === other) return true
-            if (other !is SerialDescriptorForNullable) return false
-
-            if (original != other.original) return false
+/**
+ * Returns a nullable serializer for the given serializer of non-null type.
+ */
+public fun <T : Any> KSerializer<T>.nullable(): KSerializer<T?> {
+    return NullableSerializer(this)
+}
 
-            return true
-        }
+@Deprecated(
+    message = "Deprecated in the favor of extension",
+    level = DeprecationLevel.WARNING,
+    replaceWith = ReplaceWith("actualSerializer.nullable()")
+)
+fun <T : Any> makeNullable(actualSerializer: KSerializer<T>): KSerializer<T?> {
+    return NullableSerializer(actualSerializer)
+}
 
-        override fun hashCode(): Int {
-            return original.hashCode() * 31
-        }
-    }
+internal class NullableSerializer<T : Any>(private val serializer: KSerializer<T>) : KSerializer<T?> {
```

Comment:
It can't have `internal` modifier – plugins for Native and JS use invocations of this constructor directly

Reply:
It's not the first time I bump into this.
Introduced `InternalSerializationApi`, if it looks good, let's gradually mark all plugin-used API with that (that is supposed to be `internal`, at least)

Reply:
By the way, why does plugin use both `makeNullable` and `NullableSerializer`?
Couldn't it use only `.nullable` extension?

Reply:
It's for historical reasons. I'll remove `makeNullable` invocation and replace it with `NullableSerialzier` constructor call